File: AutoPlaque-QSM-python/core/process_subject.py
"""
process_subject - orchestrate the analysis pipeline for a single subject.
"""

import os
import traceback
import numpy as np
import nibabel as nib
from pathlib import Path
from dataclasses import dataclass
from typing import Any

from algorithms.processing import process_nii_roi
from utils.extract_slice import extract_slice
from utils.mask_utils import make_whole_mask
from utils.result_table import init_result_table, update_result_table, save_results
from utils.create_nii_from_pngs import create_nii_from_pngs


def process_subject(subject_folder: str, subject_info: Any, cfg: dict) -> None:
    """
    Run the full QSM plaque detection pipeline on one subject.

    Parameters
    ----------
    subject_folder : str, absolute path to the subject directory
    subject_info   : SubjectInfo dataclass (has .qsm and .label list attributes)
    cfg            : dict with keys:
                       cfg['params']['desiredLabels']
                       cfg['params']['neiSize']
                       cfg['params']['neiOrder']
                       cfg['params']['energy']
                       cfg['params']['sparse']
                       cfg['pattern']['qsm']
    """
    params  = cfg["params"]
    pattern = cfg["pattern"]

    # 1. Paths
    qsm_filename   = subject_info.qsm[0]
    label_filename = subject_info.label[0]
    qsm_file   = str(Path(subject_folder) / qsm_filename)
    label_file = str(Path(subject_folder) / label_filename)
    output_dir = str(Path(subject_folder) / "Results")
    os.makedirs(output_dir, exist_ok=True)

    # 2. Load NIfTI volumes
    try:
        qsm_vol   = nib.load(qsm_file).get_fdata()
        label_vol = nib.load(label_file).get_fdata().astype(np.uint16)
    except Exception as e:
        print(f"Failed to read NIfTI: {e}")
        return

    _, n_slices, _ = qsm_vol.shape

    # 3. Initialise result table
    res_table = init_result_table(params["desiredLabels"])

    print(f"--> Analysing subject in: {subject_folder}")

    # 4. Slice-wise loop
    for j in range(1, n_slices + 1):
        slice_id = f"{qsm_filename.replace('.nii', '')}_slice_{j:03d}"

        frame, mask = extract_slice(qsm_vol, label_vol, j)
        sel_labels, binary_mask = make_whole_mask(mask, params["desiredLabels"])

        if sel_labels.size == 0:
            continue

        # Mask the QSM frame
        masked_frame = frame * binary_mask.astype(np.float64)

        detection_map, processed_img = process_nii_roi(
            image_frame       = masked_frame,
            slice_id          = slice_id,
            output_dir        = output_dir,
            neighborhood_size = params["neiSize"],
            do_decomposition  = True,
            roi_mask          = binary_mask,
            neighbor_order    = params["neiOrder"],
            energy_threshold  = params["energy"],
            sparse_threshold  = params["sparse"],
        )

        res_table = update_result_table(
            res_table, sel_labels, frame, mask, detection_map, processed_img
        )

    # 5. Persist results
    save_results(res_table, subject_folder, qsm_filename)
    create_nii_from_pngs(pattern["qsm"], subject_folder)

    print(f"Analysis Complete for Subject: {subject_folder}")

File: core/processSubject.m
function processSubject(subjectFolder, subjectInfo, cfg)
% PROCESSSUBJECT Orchestrates the analysis pipeline for a single subject.
%
% This function handles I/O, slice-wise iteration, and calls the core 
% detection algorithm (processNiiROI) for amyloid plaque quantification.
%
% INPUTS:
%   subjectFolder - String, path to the subject's directory.
%   subjectInfo   - Struct, contains 'qsm' and 'label' file names.
%   cfg           - Struct, global configuration with params and patterns.

    %% 1. Path Initialization
    % Construct absolute paths for NIfTI files
    qsmFile   = fullfile(subjectFolder, string(subjectInfo.qsm));
    labelFile = fullfile(subjectFolder, string(subjectInfo.label));
    outputDir = fullfile(subjectFolder, 'Results');

    % Ensure output directory exists for telemetry (PNGs, etc.)
    if ~exist(outputDir, 'dir'), mkdir(outputDir); end

    %% 2. Data Ingestion
    % Load 3D volumes from 9.4T Bruker MRI scans
    try
        qsmVol   = niftiread(qsmFile);
        labelVol = niftiread(labelFile);
    catch ME
        fprintf('Failed to read NIfTI: %s\n', ME.message);
        return;
    end

    [~, nSlices, ~] = size(qsmVol);
    
    % Initialize the quantitative results table for target ROIs
    resTable = initResultTable(cfg.params.desiredLabels);

    %% 3. Slice-wise Analysis Loop
    fprintf('--> Analyzing subject in: %s\n', subjectFolder);

    for j = 1:nSlices
        % Identifier for current slice (used for filename persistence)
        sliceID = sprintf('%s_slice_%03d', erase(string(subjectInfo.qsm), '.nii'), j);

        % Extract the current 2D frame and corresponding mask
        [frame, mask] = extractSlice(qsmVol, labelVol, j);
        
        % Filter for the desired labels (ROIs)
        [selectedROI, binaryMask] = makeWholeMask(mask, cfg.params.desiredLabels);

        % Skip slices that do not contain any target regions of interest
        if isempty(selectedROI), continue; end

        % 4. Call Core Detection Algorithm
        % Mapping global cfg.params to the processNiiROI interface
        [detectionMap, processedImg] = processNiiROI(...
            frame .* double(binaryMask), ... % Masked QSM frame
            sliceID, ...
            outputDir, ...
            cfg.params.neiSize, ...
            1, ... % doDecomposition (Hardcoded as 1 for ADDA-style detection)
            binaryMask, ...
            cfg.params.neiOrder, ...
            cfg.params.energy, ...
            cfg.params.sparse ...
        );

        % 5. Update Global Quantitative Metrics
        resTable = updateResultTable(resTable, selectedROI, ...
                                     frame, mask, detectionMap, processedImg);
    end

    %% 6. Data Persistence
    % Save aggregated results to disk and generate secondary NIfTI outputs
    saveResults(resTable, subjectFolder, string(subjectInfo.qsm));
    create_nii_from_pngs(cfg.pattern.qsm, subjectFolder);
    
    fprintf('Analysis Complete for Subject: %s\n', subjectFolder);
end